File: python/src/data_structures/stack.py
from dataclasses import dataclass

from src.data_structures.utils import printable

class Stack:
    def __init__(self):
        self.elements = []

    def push(self, element):
        self.elements.append(element)

    def pop(self):
        if self.is_empty(): return None

        return self.elements.pop()

    def peek(self):
        if self.is_empty(): return None

        return self.elements[-1]

    def is_empty(self):
        return self.size == 0

    @property
    def size(self):
        return len(self.elements)

    def __str__(self):
        string_elements = ', '.join(printable(elem) for elem in reversed(self.elements))

        return f"Stack({string_elements})"

File: ruby/lib/data_structures/stack.rb
# frozen_string_literal: true

class Stack
  def initialize
    @elements = []
  end

  def push(element)
    elements.append(element)
  end

  def pop
    return if empty?

    last_index = size - 1
    elements.delete_at(last_index)
  end

  def peek
    elements.last
  end

  def size
    elements.size
  end

  def empty?
    size == 0
  end

  def to_s
    "Stack(#{elements.join(', ')})"
  end

  private

  attr_accessor :elements
end
